### PR TITLE
pick_address: error out in dual stack mode if both addr types cannot be provided

### PR DESCRIPTION
**Pull request by rjfd**
When ms_bind_ipv4 and ms_bind_ipv6 options are enabled, if the
public_network or cluster_network settings only contain networks of a
single type (either IPv4 or IPv6) then pick_addresses function must
fail.
    
Fixes: http://tracker.ceph.com/issues/38307
    
Signed-off-by: Ricardo Dias <rdias@suse.com>

- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/common/pick_address.cc [hunk 1]
```diff
@@ -237,8 +237,16 @@ static int fill_in_one_address(
   const struct sockaddr *found = find_ip_in_subnet_list(cct, ifa, ipv, networks,
 							interfaces, numa_node);
   if (!found) {
-    lderr(cct) << "unable to find any IP address in networks '" << networks
-	       << "' interfaces '" << interfaces << "'" << dendl;
+    std::string ip_type = "";
+    if ((ipv & CEPH_PICK_ADDRESS_IPV4) && (ipv & CEPH_PICK_ADDRESS_IPV6)) {
+      ip_type = "IPv4/IPv6";
```
Comment (liewegas) [inline src/common/pick_address.cc]: s/\// or /?

**Comment by liewegas** [comment]
Hmm, I think this change is right if we think it's the right behavior.  However, I'm not sure we actually want this.  If you do
```
ms_bind_ipv4 = true
ms_bind_ipv6 = true
public_networks = 10.0.0.0/8
```
then I think that is an invalid configuration and we should error out because we can't respect the ``ms_bind_ipv6`` setting.

**Comment by liewegas** [comment]
...unless we change ``ms_bind_*`` to an enum of true/false/maybe... or want to implicitly have this behavior so that we can turn these options on by default and have ceph be dual-stack by default.  :/

**Comment by rjfd** [comment]
@liewegas 
> then I think that is an invalid configuration and we should error out because we can't respect the `ms_bind_ipv6` setting.

Yes, I also thought that that could be a solution to this problem as well. It really depends on the semantics of ms_bind_ipv*.

Not sure what's the best approach...

**Comment by rjfd** [comment]
Actually, with the commit (https://github.com/ceph/ceph/pull/26692/commits/e47ced9acdea2554c980f287832f42c81c54a6a2) that adds the information of which kind of IP address it did not found, I think error out is the most coherent solution. I'll update the PR.

**Comment by rjfd** [comment]
@liewegas I addressed your comments and updated the description of the commits and PR.

File: src/common/pick_address.cc [hunk 2]
```diff
@@ -237,8 +237,16 @@ static int fill_in_one_address(
   const struct sockaddr *found = find_ip_in_subnet_list(cct, ifa, ipv, networks,
 							interfaces, numa_node);
   if (!found) {
-    lderr(cct) << "unable to find any IP address in networks '" << networks
-	       << "' interfaces '" << interfaces << "'" << dendl;
+    std::string ip_type = "";
+    if ((ipv & CEPH_PICK_ADDRESS_IPV4) && (ipv & CEPH_PICK_ADDRESS_IPV6)) {
+      ip_type = "IPv4 or IPv6";
```
Comment (wido) [inline src/common/pick_address.cc]: But Ceph can't run dual stacked, so why would we even add this option?

IPv4 and IPv6 are mutually exclusive with Ceph

Comment (liewegas) [reply src/common/pick_address.cc]: with the infrastructure for the v2 protocol we have most of the pieces we need to run dual stack.  i expect it's a modest effort to close the gap.  this is one of the pieces

**Comment by jonesmz** [comment]
This broke my cluster when I upgraded to nautilus.

Next time, please provide some kind of explanation in the error message that the problem is not because ipv4 addresses couldn't be found, but because they couldn't be found and both types were enabled.

**Comment by hcoin** [comment]
Second to what jonesmz wrote.  The notes to upgrade a working mimic to nautilus cluster where ipv6 was enabled simply fails to start the OSD.   I think a reasonable fix would be, if ms_bind_ipv6 = true and there is no ms_bind_ipv4 assume ms_bind_ipv4 = False.

Also notice http://docs.ceph.com/docs/nautilus/rados/configuration/network-config-ref/  makes no mention of  ms_bind_ipv4

**Pull request merged**